File: 剑指offer/1，二维数组中的查找.py
"""
给定一个二维数组，其中的元素从左到右依次递增，从上到下依次递增
时间复杂度：O（m ✖ n）
空间复杂度：O（1）
"""


def find(array, target):
    if array == []:
        return False
    row = 0
    col = len(array[0]) - 1

    while row < len(array) and col >= 0:
        cur = array[row][col]

        if cur == target:
            return True
        elif cur < target:
            row += 1
        else:
            col -= 1
    return False


a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
t = 10

res = find(a, t)
print(res)

File: cpp_version/剑指offer/1，二维数组中的查找.cpp
//
// Created by Jiang,Xinyang on 2020/7/27.
//

#include <iostream>
#include <vector>

using namespace std;

bool find(vector<vector<int>>& arr, int target){
    if (arr.empty())
        return false;

    int row = 0;
    int col = (int)arr[0].size() - 1;

    while (row < arr.size() && col >= 0){
        if (target == arr[row][col])
            return true;
        else if (target < arr[row][col])
            col--;
        else
            row++;
    }
    return false;
}


int main(){
    vector<vector<int>> arr = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int target = 5;
    int target1 = 10;

    bool res1 = find(arr, target);
    cout << res1 << endl;
    bool res2 = find(arr, target1);
    cout << res2 << endl;
}
